vllm/model_executor/models/whisper.py

```diff
@@ -0,0 +1,445 @@
+from torch import nn
+import torch
+from torch import Tensor
+from typing import List, Tuple, Optional, Union
+from transformers import WhisperConfig
+from transformers.activations import GELUActivation
+from vllm.model_executor.input_metadata import InputMetadata
+from vllm.model_executor.layers.linear import (
+    LinearMethodBase,
+    RowParallelLinear,
+    ColumnParallelLinear,
+)
+from vllm.config import AudioFeaturesConfig
+from vllm.model_executor.layers.vocab_parallel_embedding import VocabParallelEmbedding
+from vllm.model_executor.layers.enc_dec_attention import EncoderAttention, DecoderAttention, CrossAttention
+from vllm.model_executor.layers.sampler import Sampler
+from vllm.model_executor.sampling_metadata import SamplingMetadata
+from vllm.model_executor.weight_utils import hf_model_weights_iterator, default_weight_loader, load_tensor_parallel_weights
+from vllm.model_executor.parallel_utils.parallel_state import get_tensor_model_parallel_world_size, get_tensor_model_parallel_rank
+
+KVCache = Tuple[torch.Tensor, torch.Tensor]
+
+
+class WhisperPositionalEmbedding(nn.Embedding):
+
+    def __init__(self, num_positions: int, embedding_dim: int):
+        super().__init__(num_positions, embedding_dim)
+
+    def forward(self, input_ids, past_key_values_length=0, position_ids=None):
+        if position_ids is None:
+            return self.weight[past_key_values_length:past_key_values_length +
+                               input_ids.shape[1]]
+        else:
+            return self.weight[position_ids]
+
+
+class WhisperAttention(nn.Module):
+
+    def __init__(
+        self,
+        config: WhisperConfig,
+        num_heads: int,
+        is_decoder: bool = False,
+        bias: bool = True,
+        is_cross: bool = False,
+        linear_method: Optional[LinearMethodBase] = None,
+    ):
+        super().__init__()
+        self.d_model = config.d_model
+        self.total_num_heads = num_heads
+        self.num_heads = num_heads // get_tensor_model_parallel_world_size()
+        self.is_decoder = is_decoder
+        self.is_cross = is_cross
+        self.key_value_proj_dim = self.d_model
+        self.head_dim = self.d_model // self.total_num_heads
+        if (self.head_dim * self.total_num_heads) != self.d_model:
+            raise ValueError(
+                f"embed_dim must be divisible by num_heads (got `embed_dim`: {self.d_model}"
+                f" and `num_heads`: {num_heads}).")
+
+        self.scaling = self.head_dim**-0.5
+
+        self.k_proj = ColumnParallelLinear(self.d_model,
```

Comment:
Can we use `QKVColumnLinear` for this?

Reply:
Or no because of the bias?

Reply:
Maybe we should extend QKVParallelLinear for this case?

Could also do this in a follow up PR

Reply:
@robertgshaw2-neuralmagic but nothing prevents us from adding `QKVParallelLinear` support for the T5 model (once again, only for encoder attention and decoder self-attention)!